# Introduction
Hello, there!
<br>
this notebook is written by Abid Ilmun Fisabil
<br>
a student at hactiv8 Full Time Data Science bootcamp program batch-006


# Problems
1-1. creating resistance and power functions<br>
1-2. creating new column named 'R/P<br>
1-3. calculating the mean for current equals to 1 Ampere and capacity equals to odd number

# Importing Dependencies

In [1]:
import pandas as pd

# Loading Dataset

In [2]:
df = pd.read_excel('NMC1-9.xlsx', engine = 'openpyxl')# importing dataset
df.head() # inspecting first 5 rows

time(s)    V_m  Current (A)  Cycle       Cap
0        0  4.098        0.997      1  2.840817
1        1  4.096        0.998      1  2.840817
2        2  4.094        0.999      1  2.840817
3        3  4.093        0.998      1  2.840817
4        4  4.092        0.999      1  2.840817

In [3]:
df.shape

(90982, 5)

we have currently a dataset that consists of 5 columns of 90982 observations

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90982 entries, 0 to 90981
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   time(s)      90982 non-null  int64  
 1   V_m          90982 non-null  float64
 2   Current (A)  90982 non-null  float64
 3   Cycle        90982 non-null  int64  
 4   Cap          90982 non-null  float64
dtypes: float64(3), int64(2)
memory usage: 3.5 MB


we can see above that no null values detected and it seems that each column is already with the right data type

In [5]:
df.describe()

time(s)           V_m   Current (A)         Cycle           Cap
count  90982.000000  90982.000000  90982.000000  90982.000000  90982.000000
mean   45490.500000      3.607666      0.999264      4.978699      2.806214
std    26264.385433      0.322777      0.001997      2.582135      0.023283
min        0.000000      2.497000      0.993000      1.000000      2.770180
25%    22745.250000      3.422000      0.997000      3.000000      2.786860
50%    45490.500000      3.636000      0.999000      5.000000      2.804878
75%    68235.750000      3.868000      1.001000      7.000000      2.823520
max    90981.000000      4.114000      1.003000      9.000000      2.840817

- the average Current measured as Ampere is at 0.9
- battery capacity is 2.8 on average
- the baterry voltage is at 3.6 on average

# Answers

## 1-1

In [6]:
def resistance(v,i):
    '''
    this function will calculate resistance by dividing v to i
    :params:v|int/float
           :i|int/float
    :return:r|int/float
    '''
    r = v/i
    return r

def power(v,i):
    '''
    this function will calculate power by multiplying v and i
    :params:v|int/float
           :i|int/float
    :return:r|int/float
    '''
    p = v*i
    return p

## 1-2

In [7]:
#initiate a list for R/P column
rp = []

#loop through dataset
for i in range(df.index[-1]+1):
    #create conditioning power function will be used for df.Cap >= 2.5 and  and resistance function for df.Cap < 2.5
    if df.Cap[i]>=2.5:
        #create a variabel stores the result from column V_m and Current(A) calculation
        pow = power(df.V_m[i],df['Current (A)'][i])
        #add pow variable as a new element for list rp
        rp.append(pow)
    elif df.Cap[i]<2.5:
        #create a variabel stores the result from column V_m and Current(A) calculation
        res = resistance(df.V_m[i],df['Current (A)'][i])
        #add res variable as a new element for list rp
        rp.append(res)

#create a new column for our df and name it R/P
df['R/P'] = rp

#inspect df to see our new column
df.head()

time(s)    V_m  Current (A)  Cycle       Cap       R/P
0        0  4.098        0.997      1  2.840817  4.085706
1        1  4.096        0.998      1  2.840817  4.087808
2        2  4.094        0.999      1  2.840817  4.089906
3        3  4.093        0.998      1  2.840817  4.084814
4        4  4.092        0.999      1  2.840817  4.087908

## 1-3

In [8]:
# filter the data frame 
cap_filtered_df = df[(df['Current (A)'] == 1) & (df.Cycle % 2 != 0)]
# calculate the average of 'Cap' column
cap_filtered_df.Cap.mean()

2.8064845553927484

above stated that the average for battery capacity is around 2.8